File: ch05/06_affine/affine/affine.py
#!/usr/bin/env python3
"""
Minimal Interpreter with Affine Types

Language syntax:
  let x = new(42)          # Create heap value (affine type)
  let y = x                # Move x to y (x consumed)
  let z = copy(x)          # Explicit copy (x not consumed)
  let r = &x               # Borrow x (x not consumed)
  print(x)                 # Use value
  drop(x)                  # Explicit drop
  x + y                    # Arithmetic

Shows:
- Affine type tracking
- Use-after-move detection
- Explicit copy vs move semantics
- Memory management
"""

from dataclasses import dataclass
from typing import Dict, Set, Optional, List, Any
from enum import Enum


# AST NODES

class Expr:
    pass

@dataclass
class Number(Expr):
    value: int

@dataclass
class Var(Expr):
    name: str

@dataclass
class New(Expr):
    """new(value) - allocate on heap (affine type)"""
    value: Expr

@dataclass
class Copy(Expr):
    """copy(var) - explicit copy"""
    var: str

@dataclass
class Borrow(Expr):
    """&var - borrow reference"""
    var: str

@dataclass
class Deref(Expr):
    """*ref - dereference"""
    ref: Expr

@dataclass
class BinOp(Expr):
    op: str  # '+', '-', '*'
    left: Expr
    right: Expr

class Stmt:
    pass

@dataclass
class Let(Stmt):
    """let x = expr"""
    var: str
    value: Expr

@dataclass
class Drop(Stmt):
    """drop(x) - explicit cleanup"""
    var: str

@dataclass
class Print(Stmt):
    """print(expr)"""
    expr: Expr

@dataclass
class ExprStmt(Stmt):
    """Expression as statement"""
    expr: Expr


# AFFINE TYPE SYSTEM

class TypeKind(Enum):
    VALUE = "value"      # Unrestricted (copyable)
    AFFINE = "affine"    # Use at most once
    REFERENCE = "ref"    # Borrowed reference

@dataclass
class Type:
    kind: TypeKind
    
    def is_affine(self):
        return self.kind == TypeKind.AFFINE
    
    def is_reference(self):
        return self.kind == TypeKind.REFERENCE

@dataclass
class VarInfo:
    """Information about a variable"""
    typ: Type
    consumed: bool = False
    consumed_at: Optional[int] = None  # Line number where consumed
    borrowed_by: Set[str] = None       # Set of references borrowing this
    
    def __post_init__(self):
        if self.borrowed_by is None:
            self.borrowed_by = set()

class AffineChecker:
    """
    This is the core of affine type checking!
    Tracks consumption state for each variable.
    """

    def __init__(self):
        self.vars: Dict[str, VarInfo] = {}
        self.line = 0
        self.errors: List[str] = []
    
    def declare(self, var: str, typ: Type):
        """Declare a new variable"""
        self.vars[var] = VarInfo(typ)
        self.log(f"DECLARE {var}: {typ.kind.value}")
    
    def consume(self, var: str) -> bool:
        """
        Mark variable as consumed (the key operation!)
        Returns True if successful, False if already consumed
        """
        if var not in self.vars:
            self.error(f"Variable '{var}' not found")
            return False
        
        info = self.vars[var]
        
        # Check if already consumed
        if info.consumed:
            self.error(f"Use after move: '{var}' was already consumed at line {info.consumed_at}")
            return False
        
        # Check if borrowed
        if len(info.borrowed_by) > 0:
            self.error(f"Cannot move '{var}' while borrowed by {info.borrowed_by}")
            return False
        
        # Consume it!
        if info.typ.is_affine():
            info.consumed = True
            info.consumed_at = self.line
            self.log(f"CONSUME {var} (affine type)")
        else:
            self.log(f"USE {var} (copyable type)")
        
        return True
    
    def borrow(self, var: str, ref_name: str) -> bool:
        """Create a borrow (reference)"""
        if var not in self.vars:
            self.error(f"Variable '{var}' not found")
            return False
        
        info = self.vars[var]
        
        if info.consumed:
            self.error(f"Cannot borrow consumed variable '{var}'")
            return False
        
        info.borrowed_by.add(ref_name)
        self.log(f"BORROW {var} by {ref_name}")
        return True
    
    def end_borrow(self, var: str, ref_name: str):
        """End a borrow"""
        if var in self.vars:
            self.vars[var].borrowed_by.discard(ref_name)
            self.log(f"END BORROW {var} by {ref_name}")
    
    def copy_var(self, var: str) -> bool:
        """Explicit copy - doesn't consume original"""
        if var not in self.vars:
            self.error(f"Variable '{var}' not found")
            return False
        
        info = self.vars[var]
        
        if info.consumed:
            self.error(f"Cannot copy consumed variable '{var}'")
            return False
        
        self.log(f"COPY {var} (original still valid)")
        return True
    
    def check_use(self, var: str) -> bool:
        """Check if variable can be used (for reading)"""
        if var not in self.vars:
            self.error(f"Variable '{var}' not found")
            return False
        
        info = self.vars[var]
        
        if info.consumed:
            self.error(f"Use after move: '{var}' was consumed at line {info.consumed_at}")
            return False
        
        self.log(f"READ {var}")
        return True
    
    def drop_var(self, var: str):
        """Explicitly drop variable"""
        if var in self.vars:
            if not self.vars[var].consumed:
                self.consume(var)
    
    def error(self, msg: str):
        self.errors.append(f"Line {self.line}: {msg}")
        print(f"  ERROR (line {self.line}): {msg}")
    
    def log(self, msg: str):
        print(f"  [{self.line}] {msg}")


# INTERPRETER WITH MEMORY MANAGEMENT

class Memory:
    """Simulated heap"""
    
    def __init__(self):
        self.heap: Dict[int, int] = {}
        self.next_addr = 1000
        self.allocations = 0
        self.deallocations = 0
    
    def allocate(self, value: int) -> int:
        """Allocate on heap, return address"""
        addr = self.next_addr
        self.heap[addr] = value
        self.next_addr += 1
        self.allocations += 1
        print(f"    ALLOC: address {addr} = {value}")
        return addr
    
    def read(self, addr: int) -> int:
        """Read from heap"""
        if addr not in self.heap:
            raise RuntimeError(f"Invalid memory access: {addr}")
        return self.heap[addr]
    
    def free(self, addr: int):
        """Free heap memory"""
        if addr in self.heap:
            value = self.heap[addr]
            del self.heap[addr]
            self.deallocations += 1
            print(f"     FREE: address {addr} (was {value})")
    
    def stats(self):
        print(f"\n Memory Stats:")
        print(f"   Allocations: {self.allocations}")
        print(f"   Deallocations: {self.deallocations}")
        print(f"   Leaked: {self.allocations - self.deallocations}")
        if len(self.heap) > 0:
            print(f"   Still allocated: {list(self.heap.keys())}")


class Interpreter:
    """Interpreter with affine type checking"""
    
    def __init__(self):
        self.checker = AffineChecker()
        self.memory = Memory()
        self.env: Dict[str, Any] = {}  # Variable -> value (addr or int)
    
    def eval_expr(self, expr: Expr) -> tuple[Any, Type]:
        """
        Evaluate expression, return (value, type)
        Key: tracks whether expression consumes its inputs!
        """
        
        if isinstance(expr, Number):
            return expr.value, Type(TypeKind.VALUE)
        
        elif isinstance(expr, Var):
            # Using a variable - check if consumed
            if not self.checker.check_use(expr.name):
                raise RuntimeError(f"Cannot use consumed variable")
            
            if expr.name not in self.env:
                raise RuntimeError(f"Variable {expr.name} not found")
            
            return self.env[expr.name], self.checker.vars[expr.name].typ
        
        elif isinstance(expr, New):
            # Allocate on heap - returns affine type!
            val, _ = self.eval_expr(expr.value)
            addr = self.memory.allocate(val)
            return addr, Type(TypeKind.AFFINE)
        
        elif isinstance(expr, Copy):
            # Explicit copy - doesn't consume
            if not self.checker.copy_var(expr.var):
                raise RuntimeError("Cannot copy")
            
            if expr.var not in self.env:
                raise RuntimeError(f"Variable {expr.var} not found")
            
            # Copy the heap value
            old_addr = self.env[expr.var]
            old_val = self.memory.read(old_addr)
            new_addr = self.memory.allocate(old_val)
            return new_addr, Type(TypeKind.AFFINE)
        
        elif isinstance(expr, Borrow):
            # Borrow - returns reference
            # Note: In real system would need lifetime tracking
            if expr.var not in self.env:
                raise RuntimeError(f"Variable {expr.var} not found")
            
            return self.env[expr.var], Type(TypeKind.REFERENCE)
        
        elif isinstance(expr, Deref):
            # Dereference
            addr, typ = self.eval_expr(expr.ref)
            if not typ.is_reference():
                raise RuntimeError("Can only deref references")
            return self.memory.read(addr), Type(TypeKind.VALUE)
        
        elif isinstance(expr, BinOp):
            # Arithmetic - consumes affine operands!
            left_val, left_type = self.eval_expr(expr.left)
            right_val, right_type = self.eval_expr(expr.right)
            
            # If operands are heap values, read them
            if left_type.is_affine():
                left_val = self.memory.read(left_val)
            if right_type.is_affine():
                right_val = self.memory.read(right_val)
            
            if expr.op == '+':
                result = left_val + right_val
            elif expr.op == '-':
                result = left_val - right_val
            elif expr.op == '*':
                result = left_val * right_val
            else:
                raise RuntimeError(f"Unknown op: {expr.op}")
            
            return result, Type(TypeKind.VALUE)
        
        raise RuntimeError(f"Unknown expr: {expr}")
    
    def exec_stmt(self, stmt: Stmt):
        """Execute statement"""
        
        if isinstance(stmt, Let):
            # Assignment - potentially consumes RHS
            val, typ = self.eval_expr(stmt.value)
            
            # If RHS is a variable with affine type, consume it
            if isinstance(stmt.value, Var):
                if self.checker.vars[stmt.value.name].typ.is_affine():
                    self.checker.consume(stmt.value.name)
            
            # Declare new variable
            self.checker.declare(stmt.var, typ)
            self.env[stmt.var] = val
        
        elif isinstance(stmt, Drop):
            # Explicit drop
            if stmt.var not in self.env:
                raise RuntimeError(f"Variable {stmt.var} not found")
            
            if self.checker.vars[stmt.var].typ.is_affine():
                addr = self.env[stmt.var]
                self.memory.free(addr)
            
            self.checker.drop_var(stmt.var)
        
        elif isinstance(stmt, Print):
            val, typ = self.eval_expr(stmt.expr)
            
            # If printing heap value, read it
            if typ.is_affine():
                val = self.memory.read(val)
            
            print(f"    OUTPUT: {val}")
        
        elif isinstance(stmt, ExprStmt):
            self.eval_expr(stmt.expr)
    
    def run(self, stmts: List[Stmt]):
        """Run program"""
        for i, stmt in enumerate(stmts):
            self.checker.line = i + 1
            print(f"\n  Line {i + 1}: {stmt}")
            
            try:
                self.exec_stmt(stmt)
            except RuntimeError as e:
                print(f"  RUNTIME ERROR: {e}")
                break
        
        # Check for memory leaks
        print("\n" + "-"*50)
        self.memory.stats()
        
        if self.checker.errors:
            print(f"\n  Found {len(self.checker.errors)} affine type errors")
        else:
            print(f"\n  No affine type errors!")


# Samples

def example_basic():
    """Basic affine types"""
    print("-"*50)
    print("EXAMPLE 1: Basic Affine Types")
    print("-"*50)
    
    program = [
        Let("x", New(Number(42))),       # x owns heap value
        Let("y", Var("x")),              # Move x to y (x consumed!)
        Print(Var("y")),                 # OK - y is valid
        # Print(Var("x")),               # Would error: x consumed!
        Drop("y"),                       # Explicit cleanup
    ]
    
    interp = Interpreter()
    interp.run(program)

def example_use_after_move():
    """Demonstrates use-after-move error"""
    print("\n" + "-"*50)
    print("EXAMPLE 2: Use After Move (ERROR)")
    print("-"*50)
    
    program = [
        Let("x", New(Number(100))),
        Let("y", Var("x")),              # x moved to y
        Print(Var("y")),                 # OK
        Print(Var("x")),                 # ERROR: x was moved!
    ]
    
    interp = Interpreter()
    interp.run(program)

def example_copy():
    """Explicit copy"""
    print("\n" + "-"*50)
    print("EXAMPLE 3: Explicit Copy")
    print("-"*50)
    
    program = [
        Let("x", New(Number(42))),
        Let("y", Copy("x")),             # Copy x (x still valid!)
        Print(Var("x")),                 # OK - x not consumed
        Print(Var("y")),                 # OK - y is separate
        Drop("x"),
        Drop("y"),
    ]
    
    interp = Interpreter()
    interp.run(program)

def example_arithmetic():
    """Arithmetic with affine types"""
    print("\n" + "-"*50)
    print("EXAMPLE 4: Arithmetic")
    print("-"*50)
    
    program = [
        Let("x", New(Number(10))),
        Let("y", New(Number(20))),
        Let("sum", BinOp("+", Var("x"), Var("y"))),
        Print(Var("sum")),
        # x and y still valid here (we only read them)
        Drop("x"),
        Drop("y"),
    ]
    
    interp = Interpreter()
    interp.run(program)

def example_no_cleanup():
    """Shows memory leak without cleanup"""
    print("\n" + "-"*50)
    print("EXAMPLE 5: Memory Leak (no drop)")
    print("-"*50)
    
    program = [
        Let("x", New(Number(42))),
        Print(Var("x")),
        # Forgot to drop x - memory leak!
    ]
    
    interp = Interpreter()
    interp.run(program)

if __name__ == "__main__":
    example_basic()
    example_use_after_move()
    example_copy()
    example_arithmetic()
    example_no_cleanup()
    
    print("\n" + "-"*50)
    print("Key Concepts Demonstrated:")
    print("-"*50)
    print("+ Affine types: values used at most once")
    print("+ Move semantics: assignment consumes source")
    print("+ Use-after-move detection at 'runtime' (would be compile-time)")
    print("+ Explicit copy vs implicit move")
    print("+ Memory management tied to ownership")
    print("+ Memory leak detection")

File: ch05/06_affine/affine/affine.c
/*
 * Affine Type System Interpreter in C
 * 
 * Demonstrates:
 * - Affine type tracking (use at most once)
 * - Move semantics vs copy semantics
 * - Use-after-move detection
 * - Explicit memory management
 * 
 * Compile: gcc -o affine affine.c -Wall -std=c99
 * Run: ./affine
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAX_VARS 100
#define MAX_HEAP 1000
#define MAX_NAME 32


// MEMORY MANAGEMENT

typedef struct {
    int addr;
    int value;
    bool allocated;
} HeapSlot;

typedef struct {
    HeapSlot slots[MAX_HEAP];
    int next_addr;
    int allocations;
    int deallocations;
} Memory;

Memory memory;

void memory_init() {
    memory.next_addr = 1000;
    memory.allocations = 0;
    memory.deallocations = 0;
    for (int i = 0; i < MAX_HEAP; i++) {
        memory.slots[i].allocated = false;
    }
}

int memory_alloc(int value) {
    // Find free slot
    for (int i = 0; i < MAX_HEAP; i++) {
        if (!memory.slots[i].allocated) {
            memory.slots[i].addr = memory.next_addr;
            memory.slots[i].value = value;
            memory.slots[i].allocated = true;
            memory.allocations++;
            
            printf("    ALLOC: address %d = %d\n", memory.next_addr, value);
            
            return memory.next_addr++;
        }
    }
    
    fprintf(stderr, "Out of memory!\n");
    exit(1);
}

int memory_read(int addr) {
    for (int i = 0; i < MAX_HEAP; i++) {
        if (memory.slots[i].allocated && memory.slots[i].addr == addr) {
            return memory.slots[i].value;
        }
    }
    
    fprintf(stderr, "Invalid memory access: %d\n", addr);
    exit(1);
}

void memory_free(int addr) {
    for (int i = 0; i < MAX_HEAP; i++) {
        if (memory.slots[i].allocated && memory.slots[i].addr == addr) {
            printf("     FREE: address %d (was %d)\n", addr, memory.slots[i].value);
            memory.slots[i].allocated = false;
            memory.deallocations++;
            return;
        }
    }
    
    fprintf(stderr, "Double free or invalid free: %d\n", addr);
}

void memory_stats() {
    printf("\n Memory Stats:\n");
    printf("   Allocations: %d\n", memory.allocations);
    printf("   Deallocations: %d\n", memory.deallocations);
    printf("   Leaked: %d\n", memory.allocations - memory.deallocations);
    
    int leaked = 0;
    printf("   Still allocated: [");
    for (int i = 0; i < MAX_HEAP; i++) {
        if (memory.slots[i].allocated) {
            if (leaked > 0) printf(", ");
            printf("%d", memory.slots[i].addr);
            leaked++;
        }
    }
    printf("]\n");
}


// AFFINE TYPE SYSTEM

typedef enum {
    TYPE_VALUE,      // Unrestricted (copyable)
    TYPE_AFFINE,     // Use at most once
    TYPE_REFERENCE   // Borrowed reference
} TypeKind;

typedef struct {
    TypeKind kind;
} Type;

typedef struct {
    char name[MAX_NAME];
    Type type;
    int value;           // The actual value (address for affine types)
    bool consumed;       // Has this been consumed?
    int consumed_at;     // Line where consumed
    bool in_use;         // Is this slot in use?
} VarInfo;

typedef struct {
    VarInfo vars[MAX_VARS];
    int count;
    int current_line;
    int error_count;
} AffineChecker;

AffineChecker checker;

void checker_init() {
    checker.count = 0;
    checker.current_line = 0;
    checker.error_count = 0;
    for (int i = 0; i < MAX_VARS; i++) {
        checker.vars[i].in_use = false;
    }
}

const char* type_name(TypeKind kind) {
    switch (kind) {
        case TYPE_VALUE: return "value";
        case TYPE_AFFINE: return "affine";
        case TYPE_REFERENCE: return "reference";
        default: return "unknown";
    }
}

VarInfo* find_var(const char* name) {
    for (int i = 0; i < checker.count; i++) {
        if (checker.vars[i].in_use && strcmp(checker.vars[i].name, name) == 0) {
            return &checker.vars[i];
        }
    }
    return NULL;
}

void checker_error(const char* msg) {
    printf("  ERROR (line %d): %s\n", checker.current_line, msg);
    checker.error_count++;
}

void checker_log(const char* msg) {
    printf("  [%d] %s\n", checker.current_line, msg);
}

void checker_declare(const char* name, Type type, int value) {
    if (checker.count >= MAX_VARS) {
        fprintf(stderr, "Too many variables!\n");
        exit(1);
    }
    
    VarInfo* var = &checker.vars[checker.count++];
    strncpy(var->name, name, MAX_NAME - 1);
    var->name[MAX_NAME - 1] = '\0';
    var->type = type;
    var->value = value;
    var->consumed = false;
    var->consumed_at = -1;
    var->in_use = true;
    
    char log_msg[256];
    snprintf(log_msg, sizeof(log_msg), "DECLARE %s: %s", name, type_name(type.kind));
    checker_log(log_msg);
}

bool checker_consume(const char* name) {
    VarInfo* var = find_var(name);
    
    if (var == NULL) {
        char msg[256];
        snprintf(msg, sizeof(msg), "Variable '%s' not found", name);
        checker_error(msg);
        return false;
    }
    
    if (var->consumed) {
        char msg[256];
        snprintf(msg, sizeof(msg), 
                 "Use after move: '%s' was already consumed at line %d", 
                 name, var->consumed_at);
        checker_error(msg);
        return false;
    }
    
    if (var->type.kind == TYPE_AFFINE) {
        var->consumed = true;
        var->consumed_at = checker.current_line;
        
        char log_msg[256];
        snprintf(log_msg, sizeof(log_msg), "CONSUME %s (affine type)", name);
        checker_log(log_msg);
    } else {
        char log_msg[256];
        snprintf(log_msg, sizeof(log_msg), "USE %s (copyable type)", name);
        checker_log(log_msg);
    }
    
    return true;
}

bool checker_check_use(const char* name) {
    VarInfo* var = find_var(name);
    
    if (var == NULL) {
        char msg[256];
        snprintf(msg, sizeof(msg), "Variable '%s' not found", name);
        checker_error(msg);
        return false;
    }
    
    if (var->consumed) {
        char msg[256];
        snprintf(msg, sizeof(msg), 
                 "Use after move: '%s' was consumed at line %d", 
                 name, var->consumed_at);
        checker_error(msg);
        return false;
    }
    
    char log_msg[256];
    snprintf(log_msg, sizeof(log_msg), "READ %s", name);
    checker_log(log_msg);
    
    return true;
}

bool checker_copy(const char* name) {
    VarInfo* var = find_var(name);
    
    if (var == NULL) {
        char msg[256];
        snprintf(msg, sizeof(msg), "Variable '%s' not found", name);
        checker_error(msg);
        return false;
    }
    
    if (var->consumed) {
        char msg[256];
        snprintf(msg, sizeof(msg), "Cannot copy consumed variable '%s'", name);
        checker_error(msg);
        return false;
    }
    
    char log_msg[256];
    snprintf(log_msg, sizeof(log_msg), "COPY %s (original still valid)", name);
    checker_log(log_msg);
    
    return true;
}

void checker_drop(const char* name) {
    VarInfo* var = find_var(name);
    if (var && !var->consumed) {
        checker_consume(name);
    }
}


// INTERPRETER OPERATIONS

// Create new heap value (affine type)
void op_new(const char* var_name, int value) {
    int addr = memory_alloc(value);
    Type type = { .kind = TYPE_AFFINE };
    checker_declare(var_name, type, addr);
}

// Move variable (consumes source)
void op_move(const char* dst, const char* src) {
    VarInfo* src_var = find_var(src);
    
    if (src_var == NULL) {
        return;
    }
    
    if (src_var->type.kind == TYPE_AFFINE) {
        if (!checker_consume(src)) {
            return;
        }
    }
    
    checker_declare(dst, src_var->type, src_var->value);
}

// Copy variable (doesn't consume source)
void op_copy(const char* dst, const char* src) {
    VarInfo* src_var = find_var(src);
    
    if (src_var == NULL || !checker_copy(src)) {
        return;
    }
    
    // Copy the heap value
    int old_addr = src_var->value;
    int old_value = memory_read(old_addr);
    int new_addr = memory_alloc(old_value);
    
    Type type = { .kind = TYPE_AFFINE };
    checker_declare(dst, type, new_addr);
}

// Print variable
void op_print(const char* var_name) {
    VarInfo* var = find_var(var_name);
    
    if (var == NULL || !checker_check_use(var_name)) {
        return;
    }
    
    int value;
    if (var->type.kind == TYPE_AFFINE) {
        value = memory_read(var->value);
    } else {
        value = var->value;
    }
    
    printf("    OUTPUT: %d\n", value);
}

// Drop variable (free memory)
void op_drop(const char* var_name) {
    VarInfo* var = find_var(var_name);
    
    if (var == NULL) {
        return;
    }
    
    if (var->type.kind == TYPE_AFFINE) {
        memory_free(var->value);
    }
    
    checker_drop(var_name);
}

// Arithmetic (reads values, doesn't consume unless affine)
void op_add(const char* result_name, const char* a_name, const char* b_name) {
    VarInfo* a = find_var(a_name);
    VarInfo* b = find_var(b_name);
    
    if (a == NULL || b == NULL) {
        return;
    }
    
    if (!checker_check_use(a_name) || !checker_check_use(b_name)) {
        return;
    }
    
    int a_val = (a->type.kind == TYPE_AFFINE) ? memory_read(a->value) : a->value;
    int b_val = (b->type.kind == TYPE_AFFINE) ? memory_read(b->value) : b->value;
    
    Type type = { .kind = TYPE_VALUE };
    checker_declare(result_name, type, a_val + b_val);
}


// EXAMPLES

void example_basic() {
    printf("\n--------------------------------------------------\n");
    printf("EXAMPLE 1: Basic Affine Types\n");
    printf("--------------------------------------------------\n");
    
    memory_init();
    checker_init();
    
    checker.current_line = 1;
    printf("\n  Line 1: let x = new(42)\n");
    op_new("x", 42);
    
    checker.current_line = 2;
    printf("\n  Line 2: let y = x (move)\n");
    op_move("y", "x");
    
    checker.current_line = 3;
    printf("\n  Line 3: print(y)\n");
    op_print("y");
    
    checker.current_line = 4;
    printf("\n  Line 4: drop(y)\n");
    op_drop("y");
    
    printf("--------------------------------------------------\n");
    memory_stats();
    
    if (checker.error_count == 0) {
        printf("\n  No affine type errors!\n");
    } else {
        printf("\n  Found %d affine type errors\n", checker.error_count);
    }
}

void example_use_after_move() {
    printf("\n--------------------------------------------------\n");
    printf("EXAMPLE 2: Use After Move (ERROR)\n");
    printf("--------------------------------------------------\n");
    
    memory_init();
    checker_init();
    
    checker.current_line = 1;
    printf("\n  Line 1: let x = new(100)\n");
    op_new("x", 100);
    
    checker.current_line = 2;
    printf("\n  Line 2: let y = x (move)\n");
    op_move("y", "x");
    
    checker.current_line = 3;
    printf("\n  Line 3: print(y)\n");
    op_print("y");
    
    checker.current_line = 4;
    printf("\n  Line 4: print(x) -- ERROR!\n");
    op_print("x");  // This will error!
    
    printf("--------------------------------------------------\n");
    memory_stats();
    
    if (checker.error_count == 0) {
        printf("\n  No affine type errors!\n");
    } else {
        printf("\n  Found %d affine type errors\n", checker.error_count);
    }
}

void example_copy() {
    printf("\n--------------------------------------------------\n");
    printf("EXAMPLE 3: Explicit Copy\n");
    printf("--------------------------------------------------\n");
    
    memory_init();
    checker_init();
    
    checker.current_line = 1;
    printf("\n  Line 1: let x = new(42)\n");
    op_new("x", 42);
    
    checker.current_line = 2;
    printf("\n  Line 2: let y = copy(x)\n");
    op_copy("y", "x");
    
    checker.current_line = 3;
    printf("\n  Line 3: print(x)\n");
    op_print("x");
    
    checker.current_line = 4;
    printf("\n  Line 4: print(y)\n");
    op_print("y");
    
    checker.current_line = 5;
    printf("\n  Line 5: drop(x)\n");
    op_drop("x");
    
    checker.current_line = 6;
    printf("\n  Line 6: drop(y)\n");
    op_drop("y");
    
    printf("--------------------------------------------------\n");
    memory_stats();
    
    if (checker.error_count == 0) {
        printf("\n  No affine type errors!\n");
    } else {
        printf("\n  Found %d affine type errors\n", checker.error_count);
    }
}

void example_arithmetic() {
    printf("\n--------------------------------------------------\n");
    printf("EXAMPLE 4: Arithmetic\n");
    printf("--------------------------------------------------\n");
    
    memory_init();
    checker_init();
    
    checker.current_line = 1;
    printf("\n  Line 1: let x = new(10)\n");
    op_new("x", 10);
    
    checker.current_line = 2;
    printf("\n  Line 2: let y = new(20)\n");
    op_new("y", 20);
    
    checker.current_line = 3;
    printf("\n  Line 3: let sum = x + y\n");
    op_add("sum", "x", "y");
    
    checker.current_line = 4;
    printf("\n  Line 4: print(sum)\n");
    op_print("sum");
    
    checker.current_line = 5;
    printf("\n  Line 5: drop(x)\n");
    op_drop("x");
    
    checker.current_line = 6;
    printf("\n  Line 6: drop(y)\n");
    op_drop("y");
    
    printf("--------------------------------------------------\n");
    memory_stats();
    
    if (checker.error_count == 0) {
        printf("\n  No affine type errors!\n");
    } else {
        printf("\n  Found %d affine type errors\n", checker.error_count);
    }
}

void example_memory_leak() {
    printf("\n--------------------------------------------------\n");
    printf("EXAMPLE 5: Memory Leak (no drop)\n");
    printf("--------------------------------------------------\n");
    
    memory_init();
    checker_init();
    
    checker.current_line = 1;
    printf("\n  Line 1: let x = new(42)\n");
    op_new("x", 42);
    
    checker.current_line = 2;
    printf("\n  Line 2: print(x)\n");
    op_print("x");
    
    printf("\n  (forgot to drop x - memory leak!)\n");
    
    printf("--------------------------------------------------\n");
    memory_stats();
    
    if (checker.error_count == 0) {
        printf("\n  No affine type errors!\n");
    } else {
        printf("\n  Found %d affine type errors\n", checker.error_count);
    }
}


// MAIN

int main() {
    example_basic();
    example_use_after_move();
    example_copy();
    example_arithmetic();
    example_memory_leak();
    
    printf("\n--------------------------------------------------\n");
    printf("Key Concepts Demonstrated:\n");
    printf("--------------------------------------------------\n");
    printf(" Affine types: values used at most once\n");
    printf(" Move semantics: assignment consumes source\n");
    printf(" Use-after-move detection\n");
    printf(" Explicit copy vs implicit move\n");
    printf(" Memory management tied to ownership\n");
    printf(" Memory leak detection\n\n");
    
    return 0;
}
